### DISCUSSION TITLE
Allocating registered buffers

### DISCUSSION THREAD

**Discussion by chris1274** [openucx/ucx#6374] [Q&A] [answered]
If I am using UCT over IB verbs, will `uct_mem_alloc` call `ibv_reg_mr` each time I call it, or is there some way to register a large amount of memory up front and just allocate subsections of it?

**Comment by yosefe** [comment, marked as answer]
`uct_mem_alloc` registers the entire buffer. In order to slice it to multiple parts, would need to implement a custom allocator to top of the buffer returned from `uct_mem_alloc`. The original memh returned from uct_mem_alloc could be used for zero-copy operation on any slice of the buffer.

**Reply by chris1274** [reply]
Thank you for the quick response. Is there any simple way to achieve the second use case (slicing the preregistered buffer with a custom allocator) with UCX or a related library?

The API pdf seems to suggest that UCS might contain memory allocators and other utilities, but there doesn't seem to be anything like that in the actual interface.

**Reply by yosefe** [reply]
UCX uses the mpool structure internally https://github.com/openucx/ucx/blob/master/src/ucs/datastruct/mpool.h, but it's not part of the official API. I'd advise implementing a similar approach in the code that uses UCT.

**Reply by chris1274** [reply]
Thanks, this is very helpful. Is UCS not meant to be used in conjunction with UCT?

**Reply by yosefe** [reply]
UCS is an internal services library that is used by UCP/UCT layers

**Reply by shamisp** [reply]
@chris1274 UCS is a unified communication "Service" layer of UCX framework (collection of libraries). It was designed to work with all the components of the UCX framework. Nevertheless, we are avoiding installing UCS headers because we don't consider the API "stable", which means the API can change in a future.
